###Part 1: NumPy – Numerical Computing

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
#Task 1: Array Creation and Operations
#1.Create the following arrays:
#A 1D array of integers from 0 to 20.
OneD_array = np.arange(0, 21)
print("\n1D array of integers from 0 to 20")
print(OneD_array)
#A 2D array of shape (4, 5) filled with random integers from 10 to 99.
TwoD_array = np.random.randint(10,100,size=(4,5))
print("\n2D array of shape (4, 5) filled with random integers from 10 to 99")
print(TwoD_array)
#A 3x3 identity matrix.
identity_matrix = np.eye(3)
print("\nA 3x3 identity matrix")
print(identity_matrix)


1D array of integers from 0 to 20
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]

2D array of shape (4, 5) filled with random integers from 10 to 99
[[39 16 44 45 28]
 [89 65 93 91 32]
 [89 94 55 26 17]
 [71 24 90 49 69]]

A 3x3 identity matrix
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
#2.Perform the following operations:
#Find the mean, median, and standard deviation of the 2D array.
mean = np.mean(TwoD_array)
median = np.median(TwoD_array)
std = np.std(TwoD_array)
print(f"Mean of 2D_array: {mean}")
print(f"Median of 2D_array: {median}")
print(f"Standard Deviation of 2D_array: {std}")

#Slice and print the second row and third column of the 2D array.
second_r = TwoD_array[1, :]
third_c = TwoD_array[:, 2]
print("\nSecond Row:", second_r)
print("Third Column:", third_c)

#Multiply two arrays element-wise.
# Reshape 1D array to match 2D array shape for element-wise multiplication
reshaped_OneD = OneD_array[:20].reshape(4, 5)
e_multiply = TwoD_array * reshaped_OneD
print("\n two arrays element-wise multiplication")
print(e_multiply)

Mean of 2D_array: 56.3
Median of 2D_array: 52.0
Standard Deviation of 2D_array: 27.25270628763316

Second Row: [89 65 93 91 32]
Third Column: [44 93 55 90]

 two arrays element-wise multiplication
[[   0   16   88  135  112]
 [ 445  390  651  728  288]
 [ 890 1034  660  338  238]
 [1065  384 1530  882 1311]]


In [ ]:
#Task 2: Broadcasting and Reshaping (15 Marks)
#1.Reshape a 1D array of size 16 into a 4x4 matrix.
One_D_arr = np.arange(16)

newshape_1d = One_D_arr.reshape(4, 4)
print("\nReshape a 1D array of size 16 into a 4x4 matrix")
print(newshape_1d)

#2.Add a 1D array [1, 2, 3, 4] to each row of the matrix using broadcasting.
add = newshape_1d + np.array([1, 2, 3, 4])
print("\nAdd a 1D array [1, 2, 3, 4] to each row of the matrix using broadcasting")
print(add)
#3.Flatten the matrix back to a 1D array.
flat = add.flatten()
print("\nFlattened Back:")
print(flat)


Reshape a 1D array of size 16 into a 4x4 matrix
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

Add a 1D array [1, 2, 3, 4] to each row of the matrix using broadcasting
[[ 1  3  5  7]
 [ 5  7  9 11]
 [ 9 11 13 15]
 [13 15 17 19]]

Flattened Back:
[ 1  3  5  7  5  7  9 11  9 11 13 15 13 15 17 19]


###Part 2: Pandas – Data Manipulation (40 Marks)

In [ ]:
#Task 3: Working with DataFrames (20 Marks)
#1.Create a DataFrame with the following data:
Data = {
    "Name": ["Alice", "Bob", "Charlie", "Diana", "Evan"],
    "Age": [25, 30, 28, 35, 40],
    "Department": ["IT", "HR", "IT", "Marketing", "HR"],
    "Salary": [50000, 45000, 72000, 58000, 60000]
}
df = pd.DataFrame(Data)
print("DataFrame:", df)

#2.Perform these operations:
#Add a new column Bonus = 10% of Salary.
df["Bonus"] = df["Salary"] * 0.10
print("\nWith Bonus Column:\n", df)

#Filter and display employees in the HR department.
hr_employees = df[df["Department"] == "HR"]
print("\nEmployees in the HR department:")
print(hr_employees)
#Calculate average salary by department.
avg_salary = df.groupby("Department")["Salary"].mean().reset_index()
print("\nAverage Salary by Department:", avg_salary)

#Save the DataFrame to a CSV file.
df.to_csv("DataFrame.csv", index=False)



DataFrame:       Name  Age Department  Salary
0    Alice   25         IT   50000
1      Bob   30         HR   45000
2  Charlie   28         IT   72000
3    Diana   35  Marketing   58000
4     Evan   40         HR   60000

With Bonus Column:
       Name  Age Department  Salary   Bonus
0    Alice   25         IT   50000  5000.0
1      Bob   30         HR   45000  4500.0
2  Charlie   28         IT   72000  7200.0
3    Diana   35  Marketing   58000  5800.0
4     Evan   40         HR   60000  6000.0

Employees in the HR department:
   Name  Age Department  Salary   Bonus
1   Bob   30         HR   45000  4500.0
4  Evan   40         HR   60000  6000.0

Average Salary by Department:   Department   Salary
0         HR  52500.0
1         IT  61000.0
2  Marketing  58000.0


In [ ]:
#Task 4: Data Cleaning & Analysis (20 Marks)
#Use any public CSV dataset (e.g., from Kaggle or UCI ML Repository)
#1.Load the dataset using pandas.read_csv()
url = "https://raw.githubusercontent.com/stedy/Machine-Learning-with-R-datasets/master/insurance.csv"
my_data = pd.read_csv(url)

In [ ]:
#2.Show:
#First 5 and last 5 rows
print("First 5 rows:")
display(my_data.head())
print("\nLast 5 rows:")
display(my_data.tail())


First 5 rows:


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520



Last 5 rows:


,age,sex,bmi,children,smoker,region,charges
1333,50,male,30.97,3,no,northwest,10600.5483
1334,18,female,31.92,0,no,northeast,2205.9808
1335,18,female,36.85,0,no,southeast,1629.8335
1336,21,female,25.80,0,no,southwest,2007.9450
1337,61,female,29.07,0,yes,northwest,29141.3603


In [ ]:
#Shape of dataset
print("\nShape of dataset:")
print(my_data.shape)



Shape of dataset:
(1338, 7)


In [ ]:
#Columns and their data types
print("\nColumns and their data types:")
print(my_data.dtypes)



Columns and their data types:
age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object


In [ ]:
#3.Handle missing values (e.g., fill, drop)
# Check which columns have missing values
my_data.isnull().sum()


,0
age,0
sex,0
bmi,0
children,0
smoker,0
region,0
charges,0


###This dataset don't have any missing values. So that, we cannot handle any kind of missing values.

In [ ]:
#4.Generate basic statistics using .describe()
print("\nBasic Statistics:")
print(my_data.describe())


Basic Statistics:
               age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010


###Part 3: Classification Evaluation with Confusion Matrix


In [ ]:
# Importing libraries
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

In [ ]:
#Task 5: Dataset-Based Classification Evaluation
# Load the Iris dataset
iris = load_iris()

# Create a DataFrame for easy handling
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

# Convert it to a binary classification problem
# 1 = Versicolor, 0 = Not Versicolor
df['binary_target'] = (df['target'] == 1).astype(int)

# Features and target
X = df[iris.feature_names]
y = df['binary_target']



In [ ]:
# 1.Split the data into 70% training and 30% testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)


In [ ]:
# 2.Train a simple Logistic Regression classifier.
model = LogisticRegression()
model.fit(X_train, y_train)


LogisticRegression()

In [ ]:
# 3.Use the model to make predictions on the test set.
y_pred = model.predict(X_test)


In [ ]:
# 4.Generate the Confusion Matrix using sklearn.metrics.confusion_matrix.
cm = confusion_matrix(y_test, y_pred)

# 5.Metrics calculation
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Display results
print("Confusion Matrix:\n", cm)
print("\nAccuracy:", round(accuracy, 3))
print("Precision:", round(precision, 3))
print("Recall:", round(recall, 3))
print("F1-Score:", round(f1, 3))


Confusion Matrix:
 [[29  3]
 [ 9  4]]

Accuracy: 0.733
Precision: 0.571
Recall: 0.308
F1-Score: 0.4


###Confusion Metrics
###True Positive (TP):
###The model correctly predicted Versicolor as Versicolor.TP = 4

###True Negative (TN):
###The model correctly predicted Not Versicolor as Not Versicolor.TN = 29

###False Positive (FP):
###The model predicted Versicolor, but it was actually Not Versicolor (a wrong positive).FP = 3

###False Negative (FN):
###The model predicted Not Versicolor, but it was actually Versicolor (a missed positive).FN = 9


###Accuracy:
###Tells how often the model was right overall.My model got about 73% of all predictions correct.

###Precision:
###Shows how many of the predicted Versicolor flowers were actually Versicolor.About 57% of the positive predictions were correct.

###Recall:
###Tells how many real Versicolor flowers the model successfully found.The recall is 30%, meaning it missed quite a few of them.

###F1-Score:
###This combines precision and recall into one score.My F1-score is 0.4, which means the model’s performance is moderate.